lib/_http_client.js

```diff
@@ -362,51 +348,53 @@ ClientRequest.prototype.abort = function abort() {
   }
   this.aborted = true;
   process.nextTick(emitAbortNT, this);
-  this.destroy();
+
+  // Compat. If we're aborting, we don't care about any more response data.
+  this.res?._dump();
+
+  const err = connResetException('aborted');
+  this.destroy(err);
+  this.res?.destroy(err);
 };
 
-ClientRequest.prototype.destroy = function destroy(err) {
-  if (this.destroyed) {
-    return this;
+ClientRequest.prototype._construct = function(callback) {
+  if (this.socket) {
+    callback();
+  } else {
+    this.once(kConstructed, callback);
   }
-  this.destroyed = true;
+};
 
-  // If we're aborting, we don't care about any more response data.
-  if (this.res) {
-    this.res._dump();
+ClientRequest.prototype._destroy = function(err, callback) {
+  if (!err && (!this.res || !this.writableFinished)) {
+    err = connResetException('socket hang up');
   }
 
   // In the event that we don't have a socket, we will pop out of
   // the request queue through handling in onSocket.
   if (this.socket) {
-    _destroy(this, this.socket, err);
-  } else if (err) {
-    this[kError] = err;
-  }
+    // TODO (ronag): Check if socket was used at all (e.g. headersSent) and
+    // re-use it in that case. `req.socket` just checks whether the socket was
+    // assigned to the request and *might* have been used.
 
-  return this;
-};
-
-function _destroy(req, socket, err) {
-  // TODO (ronag): Check if socket was used at all (e.g. headersSent) and
-  // re-use it in that case. `req.socket` just checks whether the socket was
-  // assigned to the request and *might* have been used.
-  if (socket && (!req.agent || req.socket)) {
```

Comment:
Should we keep the check on the `agent` and emit `free` instead of destroying the socket?

Reply:
I think we already do this.